File: converter.py
import csv
import re
import json
import yaml
import argparse
def read_preferences(pref_file_path, file_format):
    preferences = {}
    cell_specific_styles = {}
    user_preferences = {}
    if file_format == 'json':
        try:
            with open(pref_file_path, 'r', encoding='utf-8') as pref_file:
                data = json.load(pref_file)
                preferences.update(data.get("default", {}))
                user_preferences.update(data.get("user", {}))
                cell_specific_styles.update(data.get("cell_specific", {}))
        except json.JSONDecodeError as e:
            print(f"Error reading JSON preferences file: {e}")
            return {}
    elif file_format == 'yaml':
        try:
            with open(pref_file_path, 'r', encoding='utf-8') as pref_file:
                data = yaml.safe_load(pref_file)
                preferences.update(data.get("default", {}))
                user_preferences.update(data.get("user", {}))
                cell_specific_styles.update(data.get("cell_specific", {}))
        except yaml.YAMLError as e:
            print(f"Error reading YAML preferences file: {e}")
            return {}
    else:
        with open(pref_file_path, 'r', encoding='utf-8') as pref_file:
            mode = None
            current_specifier = None
            specific_styles = {}
            for line in pref_file:
                stripped_line = line.strip()
                if stripped_line.startswith('default:{'):
                    mode = 'default'
                    continue
                elif stripped_line.startswith('user:{'):
                    mode = 'user'
                    continue
                elif stripped_line == '}':
                    if current_specifier and 'cell_specific' in current_specifier:
                        parts = current_specifier.split('(')[1].split(')')[0].split(',')
                        if len(parts) == 2:
                            row_number = int(parts[0])
                            column_number = int(parts[1])
                            if row_number not in cell_specific_styles:
                                cell_specific_styles[row_number] = {}
                            cell_specific_styles[row_number][column_number] = specific_styles
                    current_specifier = None
                    specific_styles = {}
                    mode = None
                    continue
                if stripped_line and not stripped_line.startswith('//'):
                    if stripped_line.startswith('cell_specific'):
                        current_specifier = stripped_line.split(':')[0].strip()
                        continue
                    if current_specifier:
                        if ':' in stripped_line:
                            key, value = stripped_line.split(':', 1)
                            specific_styles[key.strip()] = value.strip().rstrip(';')
                    else:
                        if ':' in stripped_line:
                            key, value = stripped_line.split(':', 1)
                            if mode == 'default':
                                preferences[key.strip()] = value.strip().rstrip(';')
                            else:
                                user_preferences[key.strip()] = value.strip().rstrip(';')

    preferences['cell_specific'] = cell_specific_styles
    preferences.update(user_preferences)
    return preferences

def read_color_themes(theme_file_path, file_format):
    color_themes = {}

    if file_format == 'json':
        try:
            with open(theme_file_path, 'r', encoding='utf-8') as theme_file:
                color_themes = json.load(theme_file)
        except json.JSONDecodeError as e:
            print(f"Error reading JSON color themes file: {e}")
            return {}
    elif file_format == 'yaml':
        try:
            with open(theme_file_path, 'r', encoding='utf-8') as theme_file:
                color_themes = yaml.safe_load(theme_file)
        except yaml.YAMLError as e:
            print(f"Error reading YAML color themes file: {e}")
            return {}
    else:
        with open(theme_file_path, 'r', encoding='utf-8') as theme_file:
            current_theme = None
            for line in theme_file:
                stripped_line = line.strip()
                if stripped_line.endswith(':{'):
                    current_theme = stripped_line[:-2]
                    color_themes[current_theme] = {}
                elif stripped_line == '}':
                    current_theme = None
                elif current_theme and stripped_line and not stripped_line.startswith('//'):
                    if ':' in stripped_line:
                        key, value = stripped_line.split(':', 1)
                        color_themes[current_theme][key.strip()] = value.strip().rstrip(';')
    return color_themes

def read_defaults(defaults_file_path):
    defaults = {}
    with open(defaults_file_path, 'r', encoding='utf-8') as defaults_file:
        for line in defaults_file:
            stripped_line = line.strip()
            if stripped_line and not stripped_line.startswith('//'):
                key, value = stripped_line.split(':')
                defaults[key.strip()] = value.strip().rstrip(';')
    return defaults
def read_css(css_file_path):
    try:
        with open(css_file_path, 'r', encoding='utf-8') as css_file:
            return css_file.read()
    except FileNotFoundError:
        print(f"Error: CSS file {css_file_path} not found.")
        return ""
def apply_specific_styles(html_content, specific_styles, row_index, column_index):
    if row_index in specific_styles and column_index in specific_styles[row_index]:
        styles = specific_styles[row_index][column_index]
        style_string = ""
        for key, value in styles.items():
            if key == "color":
                style_string += f"background-color:{value};"
            elif key == "font":
                style_string += f"font-family:{value};"
            elif key == "font_color":
                style_string += f"color:{value};"
            elif key == "font_size":
                style_string += f"font-size:{value};"
            elif key == "border_color":
                style_string += f"border-color:{value};"
            elif key == "bold" and value.lower() == "true":
                style_string += "font-weight:bold;"
            elif key == "italics" and value.lower() == "true":
                style_string += "font-style:italic;"
            elif key == "strikethrough" and value.lower() == "true":
                style_string += "text-decoration:line-through;"
        pattern = re.compile(f'(<tr>.*?</tr>)', re.DOTALL)
        matches = pattern.findall(html_content)
        if matches and row_index < len(matches):
            row = matches[row_index]
            pattern = re.compile(f'(<td[^>]*>(?:(?!</td>).)*</td>)')
            cells = pattern.findall(row)
            if cells and column_index < len(cells):
                cell = cells[column_index]
                replacement = cell.replace('<td', f'<td style="{style_string}"')
                new_row = row.replace(cell, replacement, 1)
                html_content = html_content.replace(row, new_row, 1)
    return html_content

def csv_to_html(csv_file_path, html_file_path, preferences, color_themes, default_preferences, css_content):
    settings = {**default_preferences, **preferences}
    selected_theme = settings.get("colortheme")
    if selected_theme and selected_theme in color_themes:
        theme_settings = color_themes[selected_theme]
        settings = {**settings, **theme_settings}

    top_row_color = settings["top_row_color"]
    top_column_color = settings["top_column_color"]
    alt_color_1 = settings["alt_color_1"]
    alt_color_2 = settings["alt_color_2"]
    alt_color_3 = settings["alt_color_3"]
    alt_color_4 = settings["alt_color_4"]
    background_color = settings["background_color"]
    cell_font_name = settings["cell_font_name"]
    cell_text_color = settings["cell_text_color"]
    border_color = settings["border_color"]
    border_thickness = settings["border_thickness"]
    anti_banding = settings["anti_banding"].lower() == "true"
    title_flag = settings["title"].lower() == "true"
    title_text = settings["title_text"]
    title_color = settings["title_color"]
    website_title = settings["website_title"]
    row_banding = settings["row_banding"].lower() == "true"
    column_banding = settings["column_banding"].lower() == "true"
    cell_specific_styles = settings.get("cell_specific", {})
    switcher = settings.get("switcher", "false").lower() == "true"
    switcher_font = settings.get("switcher_font", "Arial")
    switcher_color = settings.get("switcher_color", "#000000")
    switcher_font_size = settings.get("switcher_font_size", "14px")
    if title_text == "":
        title_text = "CSV Data"
    if anti_banding:
        alt_color_1, alt_color_2 = alt_color_2, alt_color_1
    with open(csv_file_path, mode='r', newline='', encoding='utf-8') as csvfile:
        reader = csv.reader(csvfile)
        headers = next(reader)
        html_content = f'''
        <html lang="en">
        <head>
            <title>{website_title}</title>
            <style>
                :root {{
                    --background-color: {background_color};
                    --top-row-color: {top_row_color};
                    --top-column-color: {top_column_color};
                    --alt-color-1: {alt_color_1};
                    --alt-color-2: {alt_color_2};
                    --alt-color-3: {alt_color_3};
                    --alt-color-4: {alt_color_4};
                    --cell-font-name: '{cell_font_name}';
                    --cell-text-color: {cell_text_color};
                    --border-color: {border_color};
                    --border-thickness: {border_thickness};
                }}
                h1 {{
                    text-align: center;
                    color: {title_color};
                }}
                body {{
                    background-color: var(--background-color);
                    font-family: var(--cell-font-name), sans-serif;
                }}
                table {{
                    width: 100%;
                    border-collapse: collapse;
                }}
                th, td {{
                    color: var(--cell-text-color);
                    border: var(--border-thickness) solid var(--border-color);
                    padding: 8px;
                    text-align: left;
                }}
                tr:first-child {{
                    background-color: var(--top-row-color);
                }}
                td:first-child {{
                    background-color: var(--top-column-color);
                }}
                {'tr:nth-child(2n+1) td { background-color: var(--alt-color-1); }' if row_banding else ''}
                {'tr:nth-child(2n+2) td { background-color: var(--alt-color-2); }' if row_banding else ''}
                {'' if row_banding else 'tr td { background-color: var(--alt-color-1); }'}
                {''.join(['td:nth-child(2n+1) { background-color: var(--alt-color-1); }', 'td:nth-child(2n+2) { background-color: var(--alt-color-2); }']) if column_banding else ''}
                {css_content}
            </style>
            <script>
                function changeTheme(theme) {{
                    var themes = {json.dumps(color_themes)};
                    var selectedTheme = themes[theme];
                    if (selectedTheme) {{
                        document.documentElement.style.setProperty('--background-color', selectedTheme.background_color);
                        document.documentElement.style.setProperty('--top-row-color', selectedTheme.top_row_color);
                        document.documentElement.style.setProperty('--top-column-color', selectedTheme.top_column_color);
                        document.documentElement.style.setProperty('--alt-color-1', selectedTheme.alt_color_1);
                        document.documentElement.style.setProperty('--alt-color-2', selectedTheme.alt_color_2);
                        document.documentElement.style.setProperty('--cell-font-name', selectedTheme.cell_font_name);
                        document.documentElement.style.setProperty('--cell-text-color', selectedTheme.cell_text_color);
                        document.documentElement.style.setProperty('--border-color', selectedTheme.border_color);
                        document.documentElement.style.setProperty('--border-thickness', selectedTheme.border_thickness);
                    }}
                }}
            </script>
        </head>
        <body>'''

        if title_flag:
            html_content += f'<h1>{title_text}</h1>'

        if switcher:
            html_content += f'''
            <div style="text-align:center;">
                <label for="themeSwitcher" style="font-family:{switcher_font};color:{switcher_color};font-size:{switcher_font_size};">Select Color Theme:</label>
                <select id="themeSwitcher" onchange="changeTheme(this.value)" style="font-family:{switcher_font};color:{switcher_color};font-size:{switcher_font_size};">
            '''
            for theme in color_themes.keys():
                html_content += f'<option value="{theme}">{theme}</option>'
            html_content += '</select></div> <br>'

        html_content += '''
            <table>
                <thead>
                    <tr>'''
        for header in headers:
            html_content += f'<th>{header}</th>'
        
        html_content += '''
                    </tr>
                </thead>
                <tbody>
        '''
        row_index = 0
        for row in reader:
            row_index += 1
            row_html_content = '<tr>'
            for column_index, column in enumerate(row):
                cell_style = ""
                if row_banding:
                    cell_style = f"background-color:var(--alt-color-1);" if row_index % 2 == 1 else f"background-color:var(--alt-color-2);"
                if column_banding:
                    cell_style = f"background-color:var(--alt-color-1);" if column_index % 2 == 1 else f"background-color:var(--alt-color-2);"
                if row_banding and column_banding:
                    cell_style = f"background-color:var(--alt-color-1);" if (row_index + column_index) % 2 == 0 else f"background-color:var(--alt-color-2);"
                if column_index == 0:
                    row_html_content += f'<td style="background-color:var(--top-column-color); {cell_style}">{column}</td>'
                else:
                    row_html_content += f'<td style="{cell_style}">{column}</td>'
            row_html_content += '</tr>'
            row_html_content = apply_specific_styles(row_html_content, cell_specific_styles, row_index, column_index)
            html_content += row_html_content

        html_content += '''
                </tbody>
            </table>
        </body>
        </html>
        '''
    with open(html_file_path, mode='w', encoding='utf-8') as htmlfile:
        htmlfile.write(html_content)
def main():
    parser = argparse.ArgumentParser(description='Convert a CSV file to an HTML table with optional styling.')
    parser.add_argument('input_csv', nargs='?', default='csv_files/example.csv', help='Input CSV file path')
    parser.add_argument('output_html', nargs='?', default='output.html', help='Output HTML file path')
    parser.add_argument('pref_file', nargs='?', default='pref_files/pref.txt', help='Preferences file path')
    parser.add_argument('theme_file', nargs='?', default='colorthemes/colorthemes.txt', help='Color themes file path')
    parser.add_argument('--format', choices=['json', 'yaml', 'txt'], default='txt', help='Specify the format of the preferences and themes file (json, yaml, or txt)')
    parser.add_argument('--css', nargs='?', default='', help='Optional CSS file path for additional styling')
    args = parser.parse_args()
    default_preferences = read_defaults('defaults.txt')
    preferences = read_preferences(args.pref_file, args.format)
    color_themes = read_color_themes(args.theme_file, args.format)
    css_content = read_css(args.css) if args.css else ""
    csv_to_html(args.input_csv, args.output_html, preferences, color_themes, default_preferences, css_content)

if __name__ == '__main__':
    main()

File: converter.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <unordered_map>
#include <vector>
#include <string>
#include <regex>
#include <nlohmann/json.hpp>
#include <argparse/argparse.hpp>

using json = nlohmann::json;
using namespace std;

unordered_map<string, string> read_preferences(const string& file_path) {
    ifstream file(file_path);
    unordered_map<string, string> preferences;
    unordered_map<string, unordered_map<int, unordered_map<int, unordered_map<string, string>>>> cell_specific_styles;
    unordered_map<string, string> user_preferences;
    string line, mode, current_specifier;
    unordered_map<string, string> specific_styles;
    
    while (getline(file, line)) {
        string stripped_line = line;
        stripped_line.erase(remove(stripped_line.begin(), stripped_line.end(), ' '), stripped_line.end());
        if (stripped_line.starts_with("default:{")) {
            mode = "default";
            continue;
        } else if (stripped_line.starts_with("user:{")) {
            mode = "user";
            continue;
        } else if (stripped_line == "}") {
            if (current_specifier.starts_with("cell_specific")) {
                auto parts = current_specifier.substr(14, current_specifier.size() - 15);
                auto delimiter = parts.find(',');
                int row_number = stoi(parts.substr(0, delimiter));
                int column_number = stoi(parts.substr(delimiter + 1));
                cell_specific_styles[row_number][column_number] = specific_styles;
            }
            current_specifier.clear();
            specific_styles.clear();
            mode.clear();
            continue;
        }
        if (!stripped_line.empty() && !stripped_line.starts_with("//")) {
            if (stripped_line.starts_with("cell_specific")) {
                current_specifier = stripped_line.substr(0, stripped_line.find(':'));
                continue;
            }
            auto delimiter = stripped_line.find(':');
            string key = stripped_line.substr(0, delimiter);
            string value = stripped_line.substr(delimiter + 1);
            if (current_specifier.empty()) {
                if (mode == "default")
                    preferences[key] = value;
                else
                    user_preferences[key] = value;
            } else {
                specific_styles[key] = value;
            }
        }
    }
    
    for (const auto& [row, columns] : cell_specific_styles) {
        for (const auto& [column, styles] : columns) {
            string cell_key = "cell_specific(" + to_string(row) + "," + to_string(column) + ")";
            preferences[cell_key] = json(styles).dump();
        }
    }

    preferences.insert(user_preferences.begin(), user_preferences.end());
    return preferences;
}

unordered_map<string, unordered_map<string, string>> read_color_themes(const string& file_path) {
    ifstream file(file_path);
    unordered_map<string, unordered_map<string, string>> color_themes;
    string line, current_theme;
    
    while (getline(file, line)) {
        string stripped_line = line;
        stripped_line.erase(remove(stripped_line.begin(), stripped_line.end(), ' '), stripped_line.end());
        if (stripped_line.ends_with(":{")) {
            current_theme = stripped_line.substr(0, stripped_line.size() - 2);
            color_themes[current_theme] = unordered_map<string, string>();
        } else if (stripped_line == "}") {
            current_theme.clear();
        } else if (!current_theme.empty() && !stripped_line.empty() && !stripped_line.starts_with("//")) {
            auto delimiter = stripped_line.find(':');
            string key = stripped_line.substr(0, delimiter);
            string value = stripped_line.substr(delimiter + 1);
            color_themes[current_theme][key] = value;
        }
    }
    return color_themes;
}

unordered_map<string, string> read_defaults(const string& file_path) {
    ifstream file(file_path);
    unordered_map<string, string> defaults;
    string line;
    
    while (getline(file, line)) {
        string stripped_line = line;
        stripped_line.erase(remove(stripped_line.begin(), stripped_line.end(), ' '), stripped_line.end());
        if (!stripped_line.empty() && !stripped_line.starts_with("//")) {
            auto delimiter = stripped_line.find(':');
            string key = stripped_line.substr(0, delimiter);
            string value = stripped_line.substr(delimiter + 1);
            defaults[key] = value;
        }
    }
    return defaults;
}

string apply_specific_styles(const string& html_content, const unordered_map<string, string>& specific_styles, int row_index, int column_index) {
    auto cell_key = "cell_specific(" + to_string(row_index) + "," + to_string(column_index) + ")";
    if (specific_styles.find(cell_key) != specific_styles.end()) {
        json styles = json::parse(specific_styles.at(cell_key));
        string style_string;
        for (auto& [key, value] : styles.items()) {
            if (key == "color") style_string += "background-color:" + value.get<string>() + ";";
            if (key == "font") style_string += "font-family:" + value.get<string>() + ";";
            if (key == "font_color") style_string += "color:" + value.get<string>() + ";";
            if (key == "font_size") style_string += "font-size:" + value.get<string>() + ";";
            if (key == "border_color") style_string += "border-color:" + value.get<string>() + ";";
            if (key == "bold" && value == "true") style_string += "font-weight:bold;";
            if (key == "italics" && value == "true") style_string += "font-style:italic;";
            if (key == "strikethrough" && value == "true") style_string += "text-decoration:line-through;";
        }
        regex row_pattern("(<tr>.*?</tr>)");
        smatch row_match;
        if (regex_search(html_content, row_match, row_pattern)) {
            string row = row_match.str(row_index);
            regex cell_pattern("(<td[^>]*>(?:(?!</td>).)*</td>)");
            smatch cell_match;
            if (regex_search(row, cell_match, cell_pattern)) {
                string cell = cell_match.str(column_index);
                string replacement = regex_replace(cell, regex("<td"), "<td style=\"" + style_string + "\"");
                string new_row = regex_replace(row, regex(cell), replacement);
                return regex_replace(html_content, regex(row), new_row);
            }
        }
    }
    return html_content;
}

void csv_to_html(const string& csv_file_path, const string& html_file_path, const unordered_map<string, string>& preferences, const unordered_map<string, unordered_map<string, string>>& color_themes, const unordered_map<string, string>& default_preferences) {
    unordered_map<string, string> settings(default_preferences);
    settings.insert(preferences.begin(), preferences.end());

    string selected_theme = settings["colortheme"];
    if (!selected_theme.empty() && color_themes.find(selected_theme) != color_themes.end()) {
        settings.insert(color_themes.at(selected_theme).begin(), color_themes.at(selected_theme).end());
    }

    ifstream csvfile(csv_file_path);
    string line;
    getline(csvfile, line);
    istringstream header_stream(line);
    vector<string> headers;
    string header;
    while (getline(header_stream, header, ',')) headers.push_back(header);

    ofstream htmlfile(html_file_path);
    htmlfile << "<html lang=\"en\"><head><title>" << settings["website_title"] << "</title><style>:root {";
    htmlfile << "--background-color: " << settings["background_color"] << ";";
    htmlfile << "--top-row-color: " << settings["top_row_color"] << ";";
    htmlfile << "--top-column-color: " << settings["top_column_color"] << ";";
    htmlfile << "--alt-color-1: " << settings["alt_color_1"] << ";";
    htmlfile << "--alt-color-2: " << settings["alt_color_2"] << ";";
    htmlfile << "--alt-color-3: " << settings["alt_color_3"] << ";";
    htmlfile << "--alt-color-4: " << settings["alt_color_4"] << ";";
    htmlfile << "--cell-font-name: '" << settings["cell_font_name"] << "';";
    htmlfile << "--cell-text-color: " << settings["cell_text_color"] << ";";
    htmlfile << "--border-color: " << settings["border_color"] << ";";
    htmlfile << "--border-thickness: " << settings["border_thickness"] << ";";
    htmlfile << "}</style></head><body>";

    if (settings["title"] == "true") htmlfile << "<h1 style=\"text-align:center;color:" << settings["title_color"] << ";\">" << settings["title_text"] << "</h1>";

    if (settings["switcher"] == "true") {
        htmlfile << "<div style=\"text-align:center;\"><label for=\"themeSwitcher\" style=\"font-family:" << settings["switcher_font"] << ";color:" << settings["switcher_color"] << ";font-size:" << settings["switcher_font_size"] << ";\">Select Color Theme:</label>";
        htmlfile << "<select id=\"themeSwitcher\" onchange=\"changeTheme(this.value)\" style=\"font-family:" << settings["switcher_font"] << ";color:" << settings["switcher_color"] << ";font-size:" << settings["switcher_font_size"] << ";\">";
        for (const auto& [theme, _] : color_themes) htmlfile << "<option value=\"" << theme << "\">" << theme << "</option>";
        htmlfile << "</select></div><br>";
    }

    htmlfile << "<table><thead><tr>";
    for (const auto& header : headers) htmlfile << "<th>" << header << "</th>";
    htmlfile << "</tr></thead><tbody>";

    int row_index = 0;
    while (getline(csvfile, line)) {
        row_index++;
        istringstream row_stream(line);
        string cell;
        int column_index = 0;
        htmlfile << "<tr>";
        while (getline(row_stream, cell, ',')) {
            string cell_style = settings["row_banding"] == "true" ? (row_index % 2 == 1 ? "background-color:var(--alt-color-1);" : "background-color:var(--alt-color-2);") : "background-color:var(--alt-color-1);";
            if (settings["column_banding"] == "true") cell_style = column_index % 2 == 1 ? "background-color:var(--alt-color-1);" : "background-color:var(--alt-color-2);";
            if (settings["row_banding"] == "true" && settings["column_banding"] == "true") cell_style = (row_index + column_index) % 2 == 0 ? "background-color:var(--alt-color-1);" : "background-color:var(--alt-color-2);";
            if (column_index == 0)
                htmlfile << "<td style=\"background-color:var(--top-column-color);" << cell_style << "\">" << cell << "</td>";
            else
                htmlfile << "<td style=\"" << cell_style << "\">" << cell << "</td>";
            column_index++;
        }
        htmlfile << "</tr>";
    }
    htmlfile << "</tbody></table></body></html>";
}

int main(int argc, char** argv) {
    argparse::ArgumentParser program("csv_to_html");

    program.add_argument("input_csv").default_value(string("example.csv")).help("Input CSV file path");
    program.add_argument("output_html").default_value(string("output.html")).help("Output HTML file path");
    program.add_argument("pref_file").default_value(string("pref.txt")).help("Preferences file path");
    program.add_argument("theme_file").default_value(string("colorthemes.txt")).help("Color themes file path");

    try {
        program.parse_args(argc, argv);
    } catch (const std::runtime_error& err) {
        cerr << err.what() << endl;
        cerr << program;
        exit(1);
    }

    auto default_preferences = read_defaults("defaults.txt");
    auto preferences = read_preferences(program.get<string>("pref_file"));
    auto color_themes = read_color_themes(program.get<string>("theme_file"));

    csv_to_html(program.get<string>("input_csv"), program.get<string>("output_html"), preferences, color_themes, default_preferences);

    return 0;
}
